presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFilterManager.java

```diff
@@ -0,0 +1,228 @@
+            if (columnDomains.isPresent()) {
+                for (TupleDomain.ColumnDomain<ColumnHandle> columnDomain : columnDomains.get()) {
+                    KafkaColumnHandle columnHandle = (KafkaColumnHandle) columnDomain.getColumn();
+                    if (!KafkaInternalFieldDescription.isInternalColumn(columnHandle.getName())) {
+                        break;
+                    }
+                    KafkaInternalFieldDescription fieldDescription = KafkaInternalFieldDescription.forColumnName(columnHandle.getName());
+                    switch (fieldDescription) {
+                        case PARTITION_OFFSET_FIELD:
+                            offsetRanged = getRange(columnDomain.getDomain());
+                            break;
+                        case PARTITION_ID_FIELD:
+                            partitionIdsFiltered = getSingleSet(columnDomain.getDomain(), partitionIds);
+                            break;
+                        case OFFSET_TIMESTAMP_FIELD:
+                            offsetTimestampRanged = getRange(columnDomain.getDomain());
+                            break;
+                        default:
+                            break;
+                    }
+                }
+            }
+
+            if (offsetRanged.isPresent()) {
+                Range range = offsetRanged.get();
+                partitionBeginOffsets = calcTopicPartitionBeginOffsetMap(partitionBeginOffsets,
+                        (p) -> (range.getBegin() >= 0) ? Optional.of(range.getBegin()) : Optional.empty());
+                partitionEndOffsets = calcTopicPartitionEndOffsetMap(partitionEndOffsets,
+                        (p) -> (range.getEnd() >= 0) ? Optional.of(range.getEnd()) : Optional.empty());
+            }
+
+            final Set<Long> finalPartitionIdsFiltered = partitionIdsFiltered;
+            List<PartitionInfo> partitionFilteredInfos = partitionInfos.stream()
+                    .filter(partitionInfo -> finalPartitionIdsFiltered.contains((long) partitionInfo.partition()))
+                    .collect(Collectors.toList());
+            return new KafkaFilter(partitionFilteredInfos,
+                    partitionBeginOffsets,
+                    partitionEndOffsets,
+                    offsetTimestampRanged.isPresent() && offsetTimestampRanged.get().getBegin() >= 0 ? offsetTimestampRanged.get().getBegin() : null,
+                    offsetTimestampRanged.isPresent() && offsetTimestampRanged.get().getEnd() >= 0 ? offsetTimestampRanged.get().getEnd() : null);
+        }
+        else {
+            return defKafkaFilter;
+        }
+    }
+
+    public static Map<TopicPartition, Long> calcTopicPartitionBeginOffsetMap(Map<TopicPartition, Long> partitionBeginOffsets,
+                                                                             Function<TopicPartition, Optional<Long>> genBeginOffset)
```

Comment:
Perhaps you can find a more descriptive name for `genBeginOffset`, something like `computeOffset`.